### ISSUE TITLE
Consider extending `_G.string` with `vim.*` string functions

### ISSUE THREAD

**Issue by lewis6991** [neovim/neovim#23242]
### Problem

Currently we ship functions like:
- `vim.startswith()`
- `vim.endswith()`
- `vim.split()`
- `vim.gsplit()`
- `vim.pesc()`
- `vim.trim()`

Adding these to the built-in `_G.string` table allows all the above functions to be used as methods on strings. Specifically, this allows chaining.

e.g:

```lua
local s = '   hello world    --  '
print(s:trim())  -- 'hello world --'
print(s:trim():startswith('hello')) -- true
print(s:pesc()) -- '  hello world   %-%- '
```

Note: as `string` is the only type with a default metatable, we should not do this for other types (`_G.table`).

### Expected behavior

a

**Comment by clason** [comment]
I think doing this for one type only will lead to confusion. I strongly prefer a clear separation of concerns here -- what is provided by core Lua(JIT), and what is provided by Neovim "standard library".

**Comment by lewis6991** [comment]
> I think doing this for one type only will lead to confusion. 

There's only one type this can be done for. It is the only metatable provided by Lua.

> I think doing this for one type only will lead to confusion.

Does it though? Every different Lua implementation provides different builtin functions. We even provide completely different implementations for some of Lua's builtin library (e.g. `print`, `debug`, `loadfile`), so that separation is already grey.

**Comment by XeroOl** [comment]
> Does it though? It's typical for Lua implementations to provide different/addtional builtin functions (e.g. [openresty](https://github.com/openresty/luajit2#new-lua-apis), [luajit](https://luajit.org/extensions.html)). We even provide custom implementations for some of Lua's builtin library (e.g. print, debug, loadfile), so that separation is already grey and we already consider "nlua" as a thing in its own right.

No, what you are describing is not typical. There is no precedent at all here:

OpenResty's extensions are more conservative than NeoVim; they are accessed with require. If we were to add `startswith` the way OpenResty added their extensions, it would look like this:
```lua
local startswith = require("string.startswith")
if startswith("foo", "f") then print("wow!") end
```
OpenResty does not mix their additional functions with existing lua ones. They use the correct Lua mechanism for providing functions to users: Modules.

LuaJIT also is careful to provide its additional functions through require.

* LuaJIT's `bit` extensions accessed with `require("bit")`
* Same with the `ffi` and `jit` extensions

There are cases where LuaJIT has *changed* the behavior of builtin functions, but only in a very conservative backward compatible way.

Adding additional functions inside of `string` would be unprecedented, and not in the spirit of Lua.

**Comment by bfredl** [comment]
@XeroOl what you are saying is simply just false. Luajit and  Openresty's branch on top of luajit do provide functionality in other ways than required modules.

**Comment by XeroOl** [comment]
Maybe I'm bad at reading comprehension, but from the two linked pages, it looks to me like any time they expose new functions (that don't already exist in lua), it is behind a require. OpenResty and LuaJIT provide other functionality, but, as far as I can tell, never in the form of nonstandard additions standard library.

**Comment by justinmk** [comment]
> Note: as `string` is the only type with a default metatable, we should **NOT** do this for other types (`_G.table`)

Interesting argument. I generally agree that blurring the line (more) could complicate messaging/documentation/troubleshooting.

- Libraries like [penlight](https://github.com/lunarmodules/Penlight) do [similar things](https://github.com/lunarmodules/Penlight/blob/master/lua/pl/stringx.lua#L4-L5), what happens then?
    - If we have way of detecting conflict and showing a warning, that could be good enough to move forward with this. Visibility (avoiding "hidden strangeness") is the main concern imo.
- Alternative?: provide a wrapper similar to `vim.iter()`, so modules can opt-in to "enhanced string". 
    - Still a problem if they _return_ the modified string or send it as input to other modules.
